src/main/java/org/apache/maven/plugins/help/ImportedFromLocationFormatter.java

```diff
@@ -0,0 +1,125 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.maven.plugins.help;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.maven.model.Dependency;
+import org.apache.maven.model.InputLocation;
+import org.apache.maven.model.InputSource;
+import org.apache.maven.project.MavenProject;
+import org.codehaus.plexus.util.StringUtils;
+
+/**
+ * Implementation of {@link InputLocation.StringFormatter}. Enhances the default implementation with support for
+ * following "references" (caused by e.g. dependency management imports).
+ */
+public class ImportedFromLocationFormatter extends InputLocation.StringFormatter {
+    private final Method getImportedFromMethod;
+    private final MavenProject project;
+
+    public ImportedFromLocationFormatter(final Method getImportedFromMethod, final MavenProject project) {
+        this.getImportedFromMethod = getImportedFromMethod;
+        this.project = project;
+    }
+
+    @Override
+    public String toString(InputLocation location) {
+        InputSource source = location.getSource();
+
+        String s = source.getModelId(); // by default, display modelId
+
+        if (StringUtils.isBlank(s) || s.contains("[unknown-version]")) {
+            // unless it is blank or does not provide version information
+            s = source.toString();
+        }
+
+        InputLocation importedFrom = getImportedFrom(location);
+
+        StringBuilder p = new StringBuilder();
+
+        while (importedFrom != null
+                && !source.toString().equals(importedFrom.getSource().toString())) {
+            p.append(" from ").append(importedFrom.getSource().getModelId());
+            importedFrom = getImportedFrom(importedFrom);
+        }
+
+        return '}' + s + ((location.getLineNumber() >= 0) ? ", line " + location.getLineNumber() : "") + p;
+    }
+
+    protected InputLocation getImportedFrom(final InputLocation location) {
+        try {
+            InputLocation result = (InputLocation) getImportedFromMethod.invoke(location);
+
+            if (result == null && project != null) {
+                for (Dependency dependency : project.getDependencyManagement().getDependencies()) {
```

Comment:
MNG-7344 is not supposed to be limited to dependencyManagement import, but be able to work with imports done for example by Tiles Maven Plugin. Then iterating only on dependencyManagement is suspect

notice that this opens again the question of what is the purpose of trying to do something if there war no importFrom in location

Reply:
I've no experience with the maven tiles plugin. If this results in the same model (which I assume it does) it probably would still work, we would have to test this though.